File: examples/03-ConstByValueByRef/c++/main.cpp
#include <iostream>
#include <string>

void changeInt(int i);
void changeIntRef(int& i);
void changeInt(int * pI);

void changeString(std::string s);
void changeStringRef(std::string& s);
void changeString(std::string * pS);

int main(int argc, char * argv[])
{
	int i = 10;
	changeInt(i);
	std::cout << i << std::endl;

	i = 10;
	changeIntRef(i);
	std::cout << i << std::endl;

	i = 10;
	changeInt(&i);
	std::cout << i << std::endl;

	std::string s("Run");
	changeString(s);
	std::cout << s << std::endl;

	s = "Run";
	changeStringRef(s);
	std::cout << s << std::endl;

	s = "Run";
	changeString(&s);
	std::cout << s << std::endl;
}

void changeInt(int i)
{
	i = 100;
}

void changeIntRef(int& i)
{
	i = 100;
}

void changeInt(int * pI)
{
	*pI = 100;
}

void changeString(std::string s)
{
	s = " fast!";
}

void changeStringRef(std::string& s)
{
	s = s + " fast!";
}

void changeString(std::string * pS)
{
	*pS = *pS + " fast!";
}

File: examples/03-ConstByValueByRef/java/Main.java
public class Main
{
	public static void main(String args[])
	{
		Main m = new Main();
		m.go();	
	}

	private void go()
	{
		int i = 10;
		changeInt(i);
		System.out.println(i);

		MyInt mi = new MyInt(10);
		changeMyInt(mi);
		System.out.println(mi);

		String s = "Run";
		changeString(s);
		System.out.println(s);

		StringBuffer sb = new StringBuffer("Run");
		changeStringBuffer(sb);
		System.out.println(sb);
	}

	private void changeInt(int i )
	{
		i = 100;
	}

	private void changeMyInt(MyInt i)
	{
		i.setInt(100);
	}

	private void changeString(String s)
	{
		s = s + " fast!";
	}

	private void changeStringBuffer(StringBuffer sb)
	{
		sb.append(" fast!");
	}

	private static class MyInt
	{
		private int m_i;

		MyInt(int i)
		{
			setInt(i);
		}

		public int getInt()
		{
			return m_i;
		}

		public void setInt(int i)
		{
			m_i = i;
		}

		@Override
		public String toString()
		{
			return m_i+"";
		}
	}
}